syft/pkg/cataloger/catalog.go

```diff
@@ -85,6 +86,10 @@ func Catalog(resolver source.FileResolver, release *linux.Release, catalogers ..
 				allRelationships = append(allRelationships, owningRelationships...)
 			}
 
+			if filterOutPkg(p, hasEmptyName) {
+				continue
+			}
```

Comment:
I addressed items 1 and 2 with the generic cataloger and the python package cataloger. In the generic cataloger I remove all relationships From and To a dropped package. 

> So this filtering probably needs to be done at least as upstream as in the catalogers themselves, so that we never generate relationships for packages we're going to discard.
Having package and relationships dropped by the generic cataloger still a (centralized) last barrier for invalid packages.